osrelease/scanner.go

```diff
@@ -0,0 +1,206 @@
+// Package osrelease provides an "os-release" distribution scanner.
+package osrelease
+
+import (
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"runtime/trace"
+	"strings"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/internal/scanner"
+)
+
+const (
+	scannerName    = "os-release"
+	scannerVersion = "0"
+	scannerKind    = "repo"
+)
+
+const fpath = `etc/os-release`
+
+var _ scanner.DistributionScanner = (*Scanner)(nil)
+var _ scanner.VersionedScanner = (*Scanner)(nil)
+
+// Scanner implements a scanner.DistributionScanner that examines os-release
+// files, as documented at
+// https://www.freedesktop.org/software/systemd/man/os-release.html
+type Scanner struct{}
+
+// Name implements scanner.VersionedScanner.
+func (*Scanner) Name() string { return scannerName }
+
+// Version implements scanner.VersionedScanner.
+func (*Scanner) Version() string { return scannerVersion }
+
+// Kind implements scanner.VersionedScanner.
+func (*Scanner) Kind() string { return scannerKind }
+
+// Scan reports any found os-release Distribution information in the provided
+// layer.
+//
+// It's an expected outcome to return (nil, nil) when the os-release file is not
+// present in the layer.
+func (*Scanner) Scan(l *claircore.Layer) ([]*claircore.Distribution, error) {
+	ctx := context.Background()
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+
+	var rd io.ReadSeeker
+	switch {
+	case l.Bytes != nil:
+		rd = bytes.NewReader(l.Bytes)
+	case l.LocalPath != "":
+		f, err := os.Open(l.LocalPath)
+		if err != nil {
+			return nil, fmt.Errorf("osrelease: unable to open layer: %w", err)
+		}
+		defer f.Close()
+		rd = f
+	default:
+		panic("???")
+	}
+
+	d, err := readfile(ctx, rd, fpath)
+	if err != nil {
+		return nil, err
+	}
+	if d == nil {
+		return nil, nil
+	}
+	return []*claircore.Distribution{d}, nil
+}
+
+// Readfile looks for the specifed file in the tar archive provided on r.
+func readfile(ctx context.Context, r io.ReadSeeker, n string) (*claircore.Distribution, error) {
+	defer trace.StartRegion(ctx, "readfile").End()
+	if n, err := r.Seek(0, io.SeekStart); err != nil || n != 0 {
+		return nil, err
+	}
+	tr := tar.NewReader(r)
+
+Find:
+	for ctx.Err() == nil {
+		h, err := tr.Next()
+		switch err {
+		case nil:
+		case io.EOF:
+			// Hitting EOF means we didn't find anything, and as such should
+			// just fall out.
+			return nil, nil
+		default:
+			return nil, fmt.Errorf("osrelease: error reading tar: %w", err)
+		}
+
+		if h.Name != n {
+			continue
+		}
+
+		switch {
+		case h.Typeflag == tar.TypeSymlink:
+			return readfile(ctx, r, h.Linkname)
```

Comment:
h.Linkname can be a relative path when TypeSymlink is the header type giving you something like `../../path`
that will not exist in the tar bundle and must be handled. We do this here: https://github.com/quay/claircore/blob/master/filer.go#L57-L66

review that logic and see if its necessary.